**Problem Statement** : To predict the property/house price given some parameters regarding the property/house

**Approach**

1.   First we will be analysing the data via Exploratory Data Analysis to derive useful insights. This will help us later in the data preprocessing/cleaning stage.

2.   Next we will be performing the actual preprocessing/cleaning steps required in order to clean the data and convert it into a proper format for the Machine Learning model to train. This is neccessary as the Machine Learning model can only work with specific kind of data and thus data preprocessing is required to convert into that particular format!

3.   Finally we will be training the actual Machine Learning model to predict the SalePrice of a property given some input parameters!

***IMPORTING REQUIRED LIBRARIES***

In [48]:
import os
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.neighbors import LocalOutlierFactor
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.neighbors import KNeighborsRegressor

warnings.filterwarnings('ignore')

Reading the data into a Dataframe

In [49]:
data = pd.read_csv('train1.csv', index_col = 0)

Taking a deep copy of the data so as to not modify the original data

In [50]:
data1 = data.copy(deep = True)
data2 = data.copy(deep = True)

Display the first 5 rows of the dataframe

In [51]:
data1.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   

   LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
Id                                  ...                                     
1          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
2          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   
3          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
4          Lvl    AllPub    Corner  ...        0    NaN   NaN         NaN   
5          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   

   MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                             
1        0      2    2008        WD         Normal     208500  
2        0      5    2007        WD         Normal     181500  
3        0      9    2008        WD         Normal     223500  
4        0      2    2006        WD        Abnorml     140000  
5        0     12    2008        WD         Normal     250000  

[5 rows x 80 columns]

Display the last 5 rows of the dataframe

In [52]:
data1.tail()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
1456          60       RL         62.0     7917   Pave   NaN      Reg   
1457          20       RL         85.0    13175   Pave   NaN      Reg   
1458          70       RL         66.0     9042   Pave   NaN      Reg   
1459          20       RL         68.0     9717   Pave   NaN      Reg   
1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolArea PoolQC  Fence MiscFeature  \
Id                                    ...                                      
1456         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1457         Lvl    AllPub    Inside  ...        0    NaN  MnPrv         NaN   
1458         Lvl    AllPub    Inside  ...        0    NaN  GdPrv        Shed   
1459         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
1460         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   

     MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                               
1456       0      8    2007        WD         Normal     175000  
1457       0      2    2010        WD         Normal     210000  
1458    2500      5    2010        WD         Normal     266500  
1459       0      4    2010        WD         Normal     142125  
1460       0      6    2008        WD         Normal     147500  

[5 rows x 80 columns]

Display all the columns of the dataframe

In [53]:
data1.columns

Index(['MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley',
       'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope',
       'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle',
       'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle',
       'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea',
       'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC',
       'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd',
       'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond',
       'PavedDrive', 'Wo

Display the selected column only

In [56]:
data1['MSSubClass']#.head()

Id
1       60
2       20
3       60
4       70
5       60
        ..
1456    60
1457    20
1458    70
1459    20
1460    20
Name: MSSubClass, Length: 1461, dtype: int64

Display the data of multiple selected columns

In [58]:
data1[['MSSubClass', 'SaleType', 'Utilities']]

MSSubClass SaleType Utilities
Id                                 
1             60       WD    AllPub
2             20       WD    AllPub
3             60       WD    AllPub
4             70       WD    AllPub
5             60       WD    AllPub
...          ...      ...       ...
1456          60       WD    AllPub
1457          20       WD    AllPub
1458          70       WD    AllPub
1459          20       WD    AllPub
1460          20       WD    AllPub

[1461 rows x 3 columns]

Display the Index of the Dataframe

In [57]:
data1.index

Int64Index([   1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
            ...
            1451, 1452, 1453, 1454, 1455, 1456, 1457, 1458, 1459, 1460],
           dtype='int64', name='Id', length=1461)

Create a new column based on existing columns

In [59]:
data1['Age'] = data1['YrSold'] - data1['YearBuilt']

In [60]:
data1['Age']

Id
1        5
2       31
3        7
4       91
5        8
        ..
1456     8
1457    32
1458    69
1459    60
1460    43
Name: Age, Length: 1461, dtype: int64

Select particular rows using "loc" function

In [61]:
data.loc[3]

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
3           60       RL         68.0    11250   Pave   NaN      IR1   
3           60       RL         68.0    11250   Pave   NaN      IR1   

   LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
Id                                  ...                                     
3          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
3          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   

   MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                             
3        0      9    2008        WD         Normal     223500  
3        0      9    2008        WD         Normal     223500  

[2 rows x 80 columns]

Select row with Id "3" and column name "LotArea" and "Street"

In [62]:
data1.loc[3, ['LotArea', 'Street']]

LotArea Street
Id                
3     11250   Pave
3     11250   Pave

Select all rows with column name "LotArea" and "Street"

In [63]:
data1.loc[:, ['LotArea', 'Street']]

LotArea Street
Id                  
1        8450   Pave
2        9600   Pave
3       11250   Pave
4        9550   Pave
5       14260   Pave
...       ...    ...
1456     7917   Pave
1457    13175   Pave
1458     9042   Pave
1459     9717   Pave
1460     9937   Pave

[1461 rows x 2 columns]

Select rows using "iloc" function

In [64]:
data1.iloc[2, 2]

68.0

In [65]:
data1.head(6)

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   
6           50       RL         85.0    14115   Pave   NaN      IR1   

   LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
Id                                  ...                                     
1          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
2          Lvl    AllPub       FR2  ...    NaN    NaN         NaN       0   
3          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
4          Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
5          Lvl    AllPub       FR2  ...    NaN    NaN         NaN       0   
6          Lvl    AllPub    Inside  ...    NaN  MnPrv        Shed     700   

   MoSold YrSold  SaleType  SaleCondition  SalePrice  Age  
Id                                                         
1       2   2008        WD         Normal     208500    5  
2       5   2007        WD         Normal     181500   31  
3       9   2008        WD         Normal     223500    7  
4       2   2006        WD        Abnorml     140000   91  
5      12   2008        WD         Normal     250000    8  
6      10   2009        WD         Normal     143000   16  

[6 rows x 81 columns]

Conditional selection of rows using "iloc" function

In [66]:
data1[data1['LotArea'] >= 5000]

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
1             60       RL         65.0     8450   Pave   NaN      Reg   
2             20       RL         80.0     9600   Pave   NaN      Reg   
3             60       RL         68.0    11250   Pave   NaN      IR1   
4             70       RL         60.0     9550   Pave   NaN      IR1   
5             60       RL         84.0    14260   Pave   NaN      IR1   
...          ...      ...          ...      ...    ...   ...      ...   
1456          60       RL         62.0     7917   Pave   NaN      Reg   
1457          20       RL         85.0    13175   Pave   NaN      Reg   
1458          70       RL         66.0     9042   Pave   NaN      Reg   
1459          20       RL         68.0     9717   Pave   NaN      Reg   
1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
Id                                    ...                                     
1            Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
2            Lvl    AllPub       FR2  ...    NaN    NaN         NaN       0   
3            Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
4            Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
5            Lvl    AllPub       FR2  ...    NaN    NaN         NaN       0   
...          ...       ...       ...  ...    ...    ...         ...     ...   
1456         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1457         Lvl    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
1458         Lvl    AllPub    Inside  ...    NaN  GdPrv        Shed    2500   
1459         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1460         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  Age  
Id                                                           
1         2   2008        WD         Normal     208500    5  
2         5   2007        WD         Normal     181500   31  
3         9   2008        WD         Normal     223500    7  
4         2   2006        WD        Abnorml     140000   91  
5        12   2008        WD         Normal     250000    8  
...     ...    ...       ...            ...        ...  ...  
1456      8   2007        WD         Normal     175000    8  
1457      2   2010        WD         Normal     210000   32  
1458      5   2010        WD         Normal     266500   69  
1459      4   2010        WD         Normal     142125   60  
1460      6   2008        WD         Normal     147500   43  

[1319 rows x 81 columns]

In [67]:
data1[data1['LotShape'] == 'IR1']

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
3             60       RL         68.0    11250   Pave   NaN      IR1   
4             70       RL         60.0     9550   Pave   NaN      IR1   
5             60       RL         84.0    14260   Pave   NaN      IR1   
6             50       RL         85.0    14115   Pave   NaN      IR1   
8             60       RL          NaN    10382   Pave   NaN      IR1   
...          ...      ...          ...      ...    ...   ...      ...   
1430          20       RL          NaN    12546   Pave   NaN      IR1   
1432         120       RL          NaN     4928   Pave   NaN      IR1   
1434          60       RL         93.0    10261   Pave   NaN      IR1   
1441          70       RL         79.0    11526   Pave   NaN      IR1   
1447          20       RL          NaN    26142   Pave   NaN      IR1   

     LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
Id                                    ...                                     
3            Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
4            Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
5            Lvl    AllPub       FR2  ...    NaN    NaN         NaN       0   
6            Lvl    AllPub    Inside  ...    NaN  MnPrv        Shed     700   
8            Lvl    AllPub    Corner  ...    NaN    NaN        Shed     350   
...          ...       ...       ...  ...    ...    ...         ...     ...   
1430         Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
1432         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1434         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1441         Bnk    AllPub    Inside  ...    NaN    NaN         NaN       0   
1447         Lvl    AllPub   CulDSac  ...    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  Age  
Id                                                           
3         9   2008        WD         Normal     223500    7  
4         2   2006        WD        Abnorml     140000   91  
5        12   2008        WD         Normal     250000    8  
6        10   2009        WD         Normal     143000   16  
8        11   2009        WD         Normal     200000   36  
...     ...    ...       ...            ...        ...  ...  
1430      4   2007        WD         Normal     182900   26  
1432     10   2009        WD         Normal     143750   33  
1434      5   2008        WD         Normal     186500    8  
1441      9   2008        WD         Normal     191000   86  
1447      4   2010        WD         Normal     157900   48  

[487 rows x 81 columns]

Multiple conditions in "iloc" function

In [68]:
data1[(data1['LotShape'] == 'IR1') & (data1['LotArea'] >= 50000)]

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
54            20       RL         68.0    50271   Pave   NaN      IR1   
336          190       RL          NaN   164660   Grvl   NaN      IR1   
452           20       RL         62.0    70761   Pave   NaN      IR1   
458           20       RL          NaN    53227   Pave   NaN      IR1   
1397          20       RL          NaN    57200   Pave   NaN      IR1   

     LandContour Utilities LotConfig  ... PoolQC Fence MiscFeature MiscVal  \
Id                                    ...                                    
54           Low    AllPub    Inside  ...    NaN   NaN         NaN       0   
336          HLS    AllPub    Corner  ...    NaN   NaN        Shed     700   
452          Low    AllPub    Inside  ...    NaN   NaN         NaN       0   
458          Low    AllPub   CulDSac  ...    NaN   NaN         NaN       0   
1397         Bnk    AllPub    Inside  ...    NaN   NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  Age  
Id                                                           
54       11   2006        WD         Normal     385000   25  
336       8   2008        WD         Normal     228950   43  
452      12   2006        WD         Normal     280000   31  
458       3   2008        WD         Normal     256000   54  
1397      6   2010        WD         Normal     160000   62  

[5 rows x 81 columns]

Conditional selection can help weed out outliers or incorrect data

In [69]:
data1[data1['Age'] > 100]

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
186           75       RM         90.0    22950   Pave   NaN      IR2   
243           50       RM         63.0     5000   Pave   NaN      Reg   
305           75       RM         87.0    18386   Pave   NaN      Reg   
391           50       RL         50.0     8405   Pave  Grvl      Reg   
489          190       RL         60.0    10800   Pave   NaN      Reg   
521          190       RL         60.0    10800   Pave  Grvl      Reg   
584           75       RM         75.0    13500   Pave   NaN      Reg   
631           70       RM         50.0     9000   Pave  Grvl      Reg   
654           50       RM         60.0    10320   Pave  Grvl      Reg   
677           70       RM         60.0     9600   Pave  Grvl      Reg   
704          190       RM         76.0     7630   Pave   NaN      Reg   
712           50  C (all)         66.0     8712   Pave  Pave      Reg   
717           70       RM         60.0    10800   Pave  Grvl      Reg   
748           70       RM         65.0    11700   Pave  Pave      IR1   
810           75       RM         90.0     8100   Pave   NaN      Reg   
842           70       RM         60.0    10440   Pave  Grvl      Reg   
922           90       RL         67.0     8777   Pave   NaN      Reg   
946           50       RM         98.0     8820   Pave   NaN      Reg   
992           70       RM        121.0    17671   Pave  Grvl      Reg   
1063         190       RM         85.0    13600   Pave  Grvl      Reg   
1133          70       RM         90.0     9900   Pave   NaN      Reg   
1138          50       RL         54.0     6342   Pave   NaN      Reg   
1187         190       RL        107.0    10615   Pave   NaN      IR1   
1293          70       RM         60.0     6600   Pave   NaN      Reg   
1350          70       RM         50.0     5250   Pave  Pave      Reg   
1394         190       RM         60.0    10800   Pave  Pave      Reg   
1417         190       RM         60.0    11340   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
Id                                    ...                                     
186          Lvl    AllPub    Inside  ...    NaN  GdPrv         NaN       0   
243          Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
305          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
391          Lvl    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
489          Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
521          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
584          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
631          Lvl    AllPub    Corner  ...    NaN  MnPrv         NaN       0   
654          Lvl    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
677          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
704          Lvl    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
712          HLS    AllPub    Inside  ...    NaN    NaN         NaN       0   
717          Bnk    AllPub    Inside  ...    NaN    NaN         NaN       0   
748          Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
810          Lvl    AllPub    Corner  ...    NaN  MnPrv         NaN       0   
842          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
922          Lvl    AllPub    Inside  ...    NaN  GdPrv         NaN       0   
946          Lvl    AllPub    Corner  ...    NaN   MnWw         NaN       0   
992          Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
1063         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1133         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1138         Lvl    AllPub    Inside  ...    NaN    NaN         NaN    

Get statistics of the dataframe i.e. mean, median, mode, correlation, variance, standard deviation and more

In [70]:
data1.mean()

MSSubClass           57.043121
LotFrontage          70.021613
LotArea           10526.568789
OverallQual           6.108145
OverallCond           5.574949
YearBuilt          1971.417522
YearRemodAdd       1984.881588
MasVnrArea          103.399174
BsmtFinSF1          447.819986
BsmtFinSF2           46.423682
BsmtUnfSF           564.778919
TotalBsmtSF        1059.022587
1stFlrSF           1162.201916
2ndFlrSF            349.606434
LowQualFinSF          5.840520
GrLivArea          1517.648871
BsmtFullBath          0.430527
BsmtHalfBath          0.058179
FullBath              1.566735
HalfBath              0.383299
BedroomAbvGr          2.865845
KitchenAbvGr          1.047228
TotRmsAbvGrd          6.521561
Fireplaces            0.617385
GarageYrBlt        1978.597685
GarageCars            1.772758
GarageArea          474.438741
WoodDeckSF           94.457221
OpenPorchSF          46.863107
EnclosedPorch        22.270363
3SsnPorch             3.626283
ScreenPorch          14.924709
PoolArea

In [71]:
data1.median()

MSSubClass           50.0
LotFrontage          69.0
LotArea            9500.0
OverallQual           6.0
OverallCond           5.0
YearBuilt          1973.0
YearRemodAdd       1994.0
MasVnrArea            0.0
BsmtFinSF1          387.0
BsmtFinSF2            0.0
BsmtUnfSF           473.0
TotalBsmtSF         992.0
1stFlrSF           1086.0
2ndFlrSF              0.0
LowQualFinSF          0.0
GrLivArea          1466.0
BsmtFullBath          0.0
BsmtHalfBath          0.0
FullBath              2.0
HalfBath              0.0
BedroomAbvGr          3.0
KitchenAbvGr          1.0
TotRmsAbvGrd          6.0
Fireplaces            1.0
GarageYrBlt        1980.0
GarageCars            2.0
GarageArea          480.0
WoodDeckSF            0.0
OpenPorchSF          25.0
EnclosedPorch         0.0
3SsnPorch             0.0
ScreenPorch           0.0
PoolArea              0.0
MiscVal               0.0
MoSold                6.0
YrSold             2008.0
SalePrice        163000.0
Age                  35.0
dtype: float

In [84]:
data1.mode()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          20       RL         60.0     7200   Pave  Grvl      Reg   

  LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub    Inside  ...     Gd  MnPrv        Shed       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  Age  
0      6   2009        WD         Normal     140000    1  

[1 rows x 81 columns]

In [72]:
data1.std()

MSSubClass          42.399166
LotFrontage         24.264170
LotArea           9978.398335
OverallQual          1.383591
OverallCond          1.111904
YearBuilt           30.117957
YearRemodAdd        20.621815
MasVnrArea         179.807006
BsmtFinSF1         456.225704
BsmtFinSF2         161.237175
BsmtUnfSF          441.816636
TotalBsmtSF        437.345408
1stFlrSF           386.438405
2ndFlrSF           437.525242
LowQualFinSF        48.606667
GrLivArea          525.591053
BsmtFullBath         0.519614
BsmtHalfBath         0.239941
FullBath             0.551923
HalfBath             0.502972
BedroomAbvGr         0.816653
KitchenAbvGr         0.221671
TotRmsAbvGrd         1.629894
Fireplaces           0.645914
GarageYrBlt         24.666557
GarageCars           0.745101
GarageArea         213.167067
WoodDeckSF         125.298904
OpenPorchSF         66.336846
EnclosedPorch       61.769458
3SsnPorch           30.455221
ScreenPorch         55.565312
PoolArea            40.163610
MiscVal   

In [73]:
data1.var()

MSSubClass       1.797689e+03
LotFrontage      5.887499e+02
LotArea          9.956843e+07
OverallQual      1.914324e+00
OverallCond      1.236331e+00
YearBuilt        9.070913e+02
YearRemodAdd     4.252593e+02
MasVnrArea       3.233056e+04
BsmtFinSF1       2.081419e+05
BsmtFinSF2       2.599743e+04
BsmtUnfSF        1.952019e+05
TotalBsmtSF      1.912710e+05
1stFlrSF         1.493346e+05
2ndFlrSF         1.914283e+05
LowQualFinSF     2.362608e+03
GrLivArea        2.762460e+05
BsmtFullBath     2.699990e-01
BsmtHalfBath     5.757175e-02
FullBath         3.046187e-01
HalfBath         2.529812e-01
BedroomAbvGr     6.669217e-01
KitchenAbvGr     4.913786e-02
TotRmsAbvGrd     2.656555e+00
Fireplaces       4.172044e-01
GarageYrBlt      6.084390e+02
GarageCars       5.551752e-01
GarageArea       4.544020e+04
WoodDeckSF       1.569982e+04
OpenPorchSF      4.400577e+03
EnclosedPorch    3.815466e+03
3SsnPorch        9.275205e+02
ScreenPorch      3.087504e+03
PoolArea         1.613116e+03
MiscVal   

In [74]:
data1.dtypes

MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
                  ...   
YrSold             int64
SaleType          object
SaleCondition     object
SalePrice          int64
Age                int64
Length: 81, dtype: object

In [75]:
data1.skew()

MSSubClass        1.408282
LotFrontage       2.167952
LotArea          12.207001
OverallQual       0.211145
OverallCond       0.699951
YearBuilt        -0.613119
YearRemodAdd     -0.503730
MasVnrArea        2.680841
BsmtFinSF1        1.666116
BsmtFinSF2        4.260865
BsmtUnfSF         0.931311
TotalBsmtSF       1.539384
1stFlrSF          1.380760
2ndFlrSF          0.800098
LowQualFinSF      9.014533
GrLivArea         1.359763
BsmtFullBath      0.573321
BsmtHalfBath      4.073444
FullBath          0.040390
HalfBath          0.673828
BedroomAbvGr      0.205781
KitchenAbvGr      4.450652
TotRmsAbvGrd      0.677647
Fireplaces        0.641404
GarageYrBlt      -0.654648
GarageCars       -0.342022
GarageArea        0.180645
WoodDeckSF        1.536496
OpenPorchSF       2.354891
EnclosedPorch     3.058898
3SsnPorch         9.934743
ScreenPorch       4.150322
PoolArea         14.833510
MiscVal          24.475863
MoSold            0.209596
YrSold            0.097290
SalePrice         1.873260
A

In [76]:
data1.corr()

MSSubClass  LotFrontage   LotArea  OverallQual  OverallCond  \
MSSubClass       1.000000    -0.386970 -0.140204     0.029215    -0.058640   
LotFrontage     -0.386970     1.000000  0.426891     0.251937    -0.058561   
LotArea         -0.140204     0.426891  1.000000     0.105455    -0.006480   
OverallQual      0.029215     0.251937  0.105455     1.000000    -0.092090   
OverallCond     -0.058640    -0.058561 -0.006480    -0.092090     1.000000   
YearBuilt        0.022406     0.126156  0.014888     0.570123    -0.375386   
YearRemodAdd     0.036155     0.090753  0.014200     0.551546     0.067433   
MasVnrArea       0.020924     0.195539  0.105924     0.411750    -0.125485   
BsmtFinSF1      -0.070292     0.236045  0.213482     0.240979    -0.047632   
BsmtFinSF2      -0.065841     0.050986  0.111224    -0.060263     0.040780   
BsmtUnfSF       -0.140300     0.129841 -0.003005     0.304118    -0.136310   
TotalBsmtSF     -0.239334     0.393714  0.260668     0.536393    -0.172357   
1stFlrSF        -0.253094     0.457610  0.299291     0.476225    -0.143885   
2ndFlrSF         0.304190     0.080465  0.051172     0.297097     0.026541   
LowQualFinSF     0.045936     0.038619  0.004661    -0.031173     0.025549   
GrLivArea        0.071383     0.403567  0.263081     0.594576    -0.081334   
BsmtFullBath     0.005437     0.101922  0.157673     0.114302    -0.060037   
BsmtHalfBath    -0.005633    -0.005461  0.047289    -0.041660     0.123562   
FullBath         0.126832     0.198614  0.125660     0.553818    -0.195381   
HalfBath         0.175551     0.054553  0.015183     0.271094    -0.058751   
BedroomAbvGr    -0.027190     0.263531  0.120312     0.104382     0.014852   
KitchenAbvGr     0.285237    -0.010612 -0.019115    -0.181923    -0.085233   
TotRmsAbvGrd     0.036852     0.351974  0.189722     0.431469    -0.059000   
Fireplaces      -0.041339     0.263875  0.269535     0.396585    -0.022511   
GarageYrBlt      0.079530     0.072180 -0.024227     0.547025    -0.326928   
GarageCars      -0.043940     0.287997  0.154959     0.601875    -0.186938   
GarageArea      -0.103259     0.347545  0.180599     0.562494    -0.153213   
WoodDeckSF      -0.015316     0.090400  0.171538     0.241835    -0.000315   
OpenPorchSF     -0.007651     0.152810  0.084328     0.307266    -0.033467   
EnclosedPorch   -0.011567     0.005554 -0.019506    -0.107878     0.066376   
3SsnPorch       -0.043641     0.072310  0.022131     0.022725     0.020872   
ScreenPorch     -0.026597     0.041695  0.043657     0.061846     0.058315   
PoolArea         0.008024     0.206251  0.077601     0.064677    -0.001962   
MiscVal         -0.006859     0.004648  0.038359    -0.030169     0.068289   
MoSold          -0.019816     0.015889  0.002714     0.071688    -0.005177   
YrSold          -0.018077     0.004589 -0.013760    -0.028658     0.045877   
SalePrice       -0.087284     0.353502  0.263512     0.790975    -0.078669   
Age             -0.023162    -0.125763 -0.015467    -0.570446     0.376783   

               YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  \
MSSubClass      0.022406      0.036155    0.020924   -0.070292   -0.065841   
LotFrontage     0.126156      0.090753    0.195539    0.236045    0.050986   
LotArea         0.014888      0.014200    0.105924    0.213482    0.111224   
OverallQual     0.570123      0.551546    0.411750    0.240979   -0.060263   
OverallCond    -0.375386      0.067433   -0.125485   -0.047632    0.040780   
YearBuilt       1.000000      0.597640    0.319811    0.251461   -0.050106   
YearRemodAdd    0.597640      1.000000    0.185019    0.129753   -0.066910   
MasVnrArea      0.319811      0.185019    1.000000    0.270786   -0.071834   
BsmtFinSF1      0.251461      0.129753    0.270786    1.000000   -0.052076   
BsmtFinSF2     -0.050106     -0.066910   -0.071834   -0.052076    1.000000   
BsmtUnfSF       0.145547      0.177555    0.110175   -0.498136   -0.207701   
TotalBsmtSF     0.390879      0.2900

Display and remove the duplicate rows in the Dataframe

In [77]:
data1[data1.duplicated()]

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   
6           50       RL         85.0    14115   Pave   NaN      IR1   
7           20       RL         75.0    10084   Pave   NaN      Reg   
8           60       RL          NaN    10382   Pave   NaN      IR1   
9           50       RM         51.0     6120   Pave   NaN      Reg   
10         190       RL         50.0     7420   Pave   NaN      Reg   
11          20       RL         70.0    11200   Pave   NaN      Reg   
12          60       RL         85.0    11924   Pave   NaN      IR1   

   LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
Id                                  ...                                     
1          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
2          Lvl    AllPub       FR2  ...    NaN    NaN         NaN       0   
3          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
4          Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
5          Lvl    AllPub       FR2  ...    NaN    NaN         NaN       0   
6          Lvl    AllPub    Inside  ...    NaN  MnPrv        Shed     700   
7          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
8          Lvl    AllPub    Corner  ...    NaN    NaN        Shed     350   
9          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
10         Lvl    AllPub    Corner  ...    NaN    NaN         NaN       0   
11         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
12         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   

   MoSold YrSold  SaleType  SaleCondition  SalePrice  Age  
Id                                                         
1       2   2008        WD         Normal     208500    5  
2       5   2007        WD         Normal     181500   31  
3       9   2008        WD         Normal     223500    7  
4       2   2006        WD        Abnorml     140000   91  
5      12   2008        WD         Normal     250000    8  
6      10   2009        WD         Normal     143000   16  
7       8   2007        WD         Normal     307000    3  
8      11   2009        WD         Normal     200000   36  
9       4   2008        WD        Abnorml     129900   77  
10      1   2008        WD         Normal     118000   69  
11      2   2008        WD         Normal     129500   43  
12      7   2006       New        Partial     345000    1  

[12 rows x 81 columns]

In [78]:
data1.drop_duplicates(keep = 'first', inplace = True)

In [79]:
data1[data1.duplicated()]

Empty DataFrame
Columns: [MSSubClass, MSZoning, LotFrontage, LotArea, Street, Alley, LotShape, LandContour, Utilities, LotConfig, LandSlope, Neighborhood, Condition1, Condition2, BldgType, HouseStyle, OverallQual, OverallCond, YearBuilt, YearRemodAdd, RoofStyle, RoofMatl, Exterior1st, Exterior2nd, MasVnrType, MasVnrArea, ExterQual, ExterCond, Foundation, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinSF1, BsmtFinType2, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, Heating, HeatingQC, CentralAir, Electrical, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, KitchenQual, TotRmsAbvGrd, Functional, Fireplaces, FireplaceQu, GarageType, GarageYrBlt, GarageFinish, GarageCars, GarageArea, GarageQual, GarageCond, PavedDrive, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, PoolQC, Fence, MiscFeature, MiscVal, MoSold, YrSold, SaleType, SaleCondition, SalePrice, Age]
Index: []

[0 rows x 81 columns]

Sort the dataframe in ascending or descending order based on a particular column

In [80]:
low_to_high_price = data1.sort_values('SalePrice', ascending = True)

In [81]:
low_to_high_price.head(15)

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
496           30  C (all)         60.0     7879   Pave   NaN      Reg   
917           20  C (all)         50.0     9000   Pave   NaN      Reg   
969           50       RM         50.0     5925   Pave   NaN      Reg   
534           20       RL         50.0     5000   Pave   NaN      Reg   
31            70  C (all)         50.0     8500   Pave  Pave      Reg   
711           30       RL         56.0     4130   Pave   NaN      IR1   
1338          30       RM        153.0     4118   Pave  Grvl      IR1   
1326          30       RM         40.0     3636   Pave   NaN      Reg   
706          190       RM         70.0     5600   Pave   NaN      Reg   
813           20  C (all)         66.0     8712   Grvl   NaN      Reg   
1381          30       RL         45.0     8212   Pave  Grvl      Reg   
1101          30       RL         60.0     8400   Pave   NaN      Reg   
411           20       RL         68.0     9571   Pave   NaN      Reg   
637           30       RM         51.0     6120   Pave   NaN      Reg   
376           30       RL          NaN    10020   Pave   NaN      IR1   

     LandContour Utilities LotConfig  ... PoolQC  Fence MiscFeature MiscVal  \
Id                                    ...                                     
496          Lvl    AllPub    Inside  ...    NaN   GdWo         NaN       0   
917          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
969          Lvl    AllPub    Inside  ...    NaN   GdWo         NaN       0   
534          Low    AllPub    Inside  ...    NaN    NaN         NaN       0   
31           Lvl    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
711          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1338         Bnk    AllPub    Corner  ...    NaN    NaN         NaN       0   
1326         Lvl    AllPub    Inside  ...    NaN  MnPrv         NaN       0   
706          Lvl    AllPub    Inside  ...    NaN    NaN        Othr    3500   
813          Bnk    AllPub    Inside  ...    NaN    NaN        Shed      54   
1381         Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
1101         Bnk    AllPub    Inside  ...    NaN    NaN         NaN       0   
411          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
637          Lvl    AllPub    Inside  ...    NaN    NaN         NaN       0   
376          Low    AllPub    Inside  ...    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  Age  
Id                                                           
496      11   2009        WD        Abnorml      34900   89  
917      10   2006        WD        Abnorml      35311   57  
969       5   2009        WD        Abnorml      37900   99  
534       1   2007        WD         Normal      39300   61  
31        7   2008        WD         Normal      40000   88  
711       7   2008        WD         Normal      52000   73  
1338      3   2006        WD         Normal      52500   65  
1326      1   2008        WD         Normal      55000   86  
706       7   2010        WD         Normal      55000   80  
813       6   2010        WD         Alloca      55993   58  
1381      6   2010        WD         Normal      58500   96  
1101      1   2009        WD         Normal      60000   89  
411       6   2009       COD        Abnorml      60000   51  
637       1   2009     ConLw         Normal      60000   73  
376       3   2009        WD         Normal      61000   87  

[15 rows x 81 columns]

Grouby data based on specific columns and statitical parameter

In [82]:
data1.groupby(['LotShape']).mean()#median/sum

MSSubClass  LotFrontage       LotArea  OverallQual  OverallCond  \
LotShape                                                                    
IR1        50.322245    76.034921  11905.686071     6.436590     5.559252   
IR2        52.073171    76.500000  23733.658537     6.731707     5.560976   
IR3        41.000000   138.428571  41338.200000     6.800000     4.800000   
Reg        60.926936    67.007109   8878.805889     5.892039     5.594329   

            YearBuilt  YearRemodAdd  MasVnrArea   BsmtFinSF1  BsmtFinSF2  ...  \
LotShape                                                                  ...   
IR1       1980.020790   1988.650728  130.434238   523.790021   52.669439  ...   
IR2       1985.268293   1996.024390  126.243902   468.146341  100.365854  ...   
IR3       1987.800000   1995.800000   79.600000  1114.100000   82.000000  ...   
Reg       1966.054526   1982.312977   88.220637   396.107961   40.920393  ...   

          OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch   PoolArea  \
LotShape                                                                  
IR1         52.336798      16.550936   4.769231    18.259875   3.767152   
IR2         71.170732       2.268293   7.073171    34.634146   0.000000   
IR3         64.600000       0.000000   0.000000    14.700000  48.000000   
Reg         42.738277      25.930207   2.610687    12.491821   1.893130   

            MiscVal    MoSold       YrSold      SalePrice        Age  
LotShape                                                              
IR1       74.074844  6.471933  2007.733888  206279.326403  27.713098  
IR2       23.902439  6.365854  2007.853659  239833.365854  22.585366  
IR3        0.000000  6.200000  2007.400000  216036.500000  19.600000  
Reg       27.954198  6.254089  2007.850600  164960.461287  41.796074  

[4 rows x 38 columns]

In [83]:
data1.groupby(['LotShape', 'Street']).mean()#median/sum

MSSubClass  LotFrontage        LotArea  OverallQual  \
LotShape Street                                                        
IR1      Grvl     190.00000          NaN  164660.000000     5.000000   
         Pave      50.03125    76.034921   11587.447917     6.439583   
IR2      Grvl      90.00000   110.000000    8472.000000     5.000000   
         Pave      51.12500    75.160000   24115.200000     6.775000   
IR3      Pave      41.00000   138.428571   41338.200000     6.800000   
Reg      Grvl      40.00000    79.250000   18421.500000     4.750000   
         Pave      61.01862    66.948810    8836.997809     5.897043   

                 OverallCond    YearBuilt  YearRemodAdd  MasVnrArea  \
LotShape Street                                                       
IR1      Grvl       6.000000  1965.000000   1965.000000    0.000000   
         Pave       5.558333  1980.052083   1988.700000  130.707113   
IR2      Grvl       5.000000  1963.000000   1963.000000    0.000000   
         Pave       5.575000  1985.825000   1996.850000  129.400000   
IR3      Pave       4.800000  1987.800000   1995.800000   79.600000   
Reg      Grvl       4.500000  1960.000000   1963.750000   82.500000   
         Pave       5.599124  1966.081051   1982.394304   88.245865   

                  BsmtFinSF1  BsmtFinSF2  ...  OpenPorchSF  EnclosedPorch  \
LotShape Street                           ...                               
IR1      Grvl    1249.000000  147.000000  ...     0.000000       0.000000   
         Pave     522.279167   52.472917  ...    52.445833      16.585417   
IR2      Grvl     104.000000  712.000000  ...     0.000000       0.000000   
         Pave     477.250000   85.075000  ...    72.950000       2.325000   
IR3      Pave    1114.100000   82.000000  ...    64.600000       0.000000   
Reg      Grvl     493.750000    0.000000  ...    78.750000       0.000000   
         Pave     395.680175   41.099671  ...    42.580504      26.043812   

                 3SsnPorch  ScreenPorch   PoolArea     MiscVal    MoSold  \
LotShape Street                                                            
IR1      Grvl     0.000000     0.000000   0.000000  700.000000  8.000000   
         Pave     4.779167    18.297917   3.775000   72.770833  6.468750   
IR2      Grvl     0.000000     0.000000   0.000000    0.000000  5.000000   
         Pave     7.250000    35.500000   0.000000   24.500000  6.400000   
IR3      Pave     0.000000    14.700000  48.000000    0.000000  6.200000   
Reg      Grvl     0.000000    65.750000   0.000000  153.500000  6.000000   
         Pave     2.622125    12.258488   1.901424   27.404162  6.255203   

                      YrSold      SalePrice        Age  
LotShape Street                                         
IR1      Grvl    2008.000000  228950.000000  43.000000  
         Pave    2007.733333  206232.095833  27.681250  
IR2      Grvl    2010.000000  110000.000000  47.000000  
         Pave    2007.800000  243079.200000  21.975000  
IR3      Pave    2007.400000  216036.500000  19.600000  
Reg      Grvl    2008.000000  110548.250000  48.000000  
         Pave    2007.849945  165198.849945  41.768894  

[7 rows x 38 columns]

Get general info of the dataframe i.e. data types, non null values, statitical measures and more with just a couple lines of code!

In [83]:
data1.info()

In [83]:
data1.describe()

In [32]:
data1.isnull().sum()

MSSubClass         0
MSZoning           0
LotFrontage      257
LotArea            0
Street             0
                ... 
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Age                0
Length: 81, dtype: int64

In [19]:
num = data1.select_dtypes(include = 'number')
cat = data1.select_dtypes(include = 'object')

In [ ]:
for i in num:
    num.boxplot(column = i, patch_artist = True, notch ='True')
    plt.ylabel(i)
    plt.show()

In [ ]:
sns.set_style('whitegrid')
for j in num:
    sns.distplot(data1[j], kde = True, color = 'red')
    plt.show()

In [ ]:
sns.scatterplot(x = data1['GarageArea'], y = data1['SalePrice'], palette='pastel')

In [ ]:
sns.scatterplot(x = data1['YrSold'], y = data1['SalePrice'], palette='pastel')

In [ ]:
for i in cat:
  sns.countplot(x = data1[i], palette = "Spectral")
  plt.show()

In [ ]:
sns.barplot(x = data1['LotShape'], y = data1['SalePrice'])

In [ ]:
sns.barplot(x = data1['BldgType'], y = data1['SalePrice'], ci = 0)

In [ ]:
x = data1.drop(['SalePrice'], axis = 1)
y = data1['SalePrice']

In [ ]:
train_x, test_x, train_y, test_y = train_test_split(x, y, random_state = 69)

In [ ]:
train_num = train_x.select_dtypes(include = 'number')
train_cat = train_x.select_dtypes(include = 'object')

test_num = test_x.select_dtypes(include = 'number')
test_cat = test_x.select_dtypes(include = 'object')

In [ ]:
print('Missing values before imputation \n', train_cat.isnull().sum())
train_cat.fillna(train_cat.mode().loc[0], inplace = True)
print('\n')
print('Missing values after imputation \n', cat.isnull().sum())

Missing values before imputation 
 MSZoning            0
Street              0
Alley            1030
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          6
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           26
BsmtCond           26
BsmtExposure       27
BsmtFinType1       26
BsmtFinType2       27
Heating             0
HeatingQC           0
CentralAir          0
Electrical          0
KitchenQual         0
Functional          0
FireplaceQu       517
GarageType         62
GarageFinish       62
GarageQual         62
GarageCond         62
PavedDrive          0
PoolQC           1090
Fence             884
MiscFeature      1059
SaleType            0
SaleCondition       0
dtype: int64


Miss

In [ ]:
print('Missing values before imputation \n', train_num.isnull().sum())
train_num.fillna(train_num.median(), inplace = True)
print('\n')
print('Missing values after imputation \n', train_num.isnull().sum())

Missing values before imputation 
 MSSubClass         0
LotFrontage      188
LotArea            0
OverallQual        0
OverallCond        0
YearBuilt          0
YearRemodAdd       0
MasVnrArea         6
BsmtFinSF1         0
BsmtFinSF2         0
BsmtUnfSF          0
TotalBsmtSF        0
1stFlrSF           0
2ndFlrSF           0
LowQualFinSF       0
GrLivArea          0
BsmtFullBath       0
BsmtHalfBath       0
FullBath           0
HalfBath           0
BedroomAbvGr       0
KitchenAbvGr       0
TotRmsAbvGrd       0
Fireplaces         0
GarageYrBlt       62
GarageCars         0
GarageArea         0
WoodDeckSF         0
OpenPorchSF        0
EnclosedPorch      0
3SsnPorch          0
ScreenPorch        0
PoolArea           0
MiscVal            0
MoSold             0
YrSold             0
dtype: int64


Missing values after imputation 
 MSSubClass       0
LotFrontage      0
LotArea          0
OverallQual      0
OverallCond      0
YearBuilt        0
YearRemodAdd     0
MasVnrArea       0
BsmtFinS

In [ ]:
print('Missing values before imputation \n', test_cat.isnull().sum())
test_cat.fillna(train_cat.mode().loc[0], inplace = True)
print('\n')
print('Missing values after imputation \n', test_cat.isnull().sum())

Missing values before imputation 
 MSZoning           0
Street             0
Alley            339
LotShape           0
LandContour        0
Utilities          0
LotConfig          0
LandSlope          0
Neighborhood       0
Condition1         0
Condition2         0
BldgType           0
HouseStyle         0
RoofStyle          0
RoofMatl           0
Exterior1st        0
Exterior2nd        0
MasVnrType         2
ExterQual          0
ExterCond          0
Foundation         0
BsmtQual          11
BsmtCond          11
BsmtExposure      11
BsmtFinType1      11
BsmtFinType2      11
Heating            0
HeatingQC          0
CentralAir         0
Electrical         1
KitchenQual        0
Functional         0
FireplaceQu      173
GarageType        19
GarageFinish      19
GarageQual        19
GarageCond        19
PavedDrive         0
PoolQC           363
Fence            295
MiscFeature      347
SaleType           0
SaleCondition      0
dtype: int64


Missing values after imputation 
 MSZoning     

In [ ]:
print('Missing values before imputation \n', test_num.isnull().sum())
test_num.fillna(train_num.median(), inplace = True)
print('\n')
print('Missing values after imputation \n', test_num.isnull().sum())

Missing values before imputation 
 MSSubClass        0
LotFrontage      71
LotArea           0
OverallQual       0
OverallCond       0
YearBuilt         0
YearRemodAdd      0
MasVnrArea        2
BsmtFinSF1        0
BsmtFinSF2        0
BsmtUnfSF         0
TotalBsmtSF       0
1stFlrSF          0
2ndFlrSF          0
LowQualFinSF      0
GrLivArea         0
BsmtFullBath      0
BsmtHalfBath      0
FullBath          0
HalfBath          0
BedroomAbvGr      0
KitchenAbvGr      0
TotRmsAbvGrd      0
Fireplaces        0
GarageYrBlt      19
GarageCars        0
GarageArea        0
WoodDeckSF        0
OpenPorchSF       0
EnclosedPorch     0
3SsnPorch         0
ScreenPorch       0
PoolArea          0
MiscVal           0
MoSold            0
YrSold            0
dtype: int64


Missing values after imputation 
 MSSubClass       0
LotFrontage      0
LotArea          0
OverallQual      0
OverallCond      0
YearBuilt        0
YearRemodAdd     0
MasVnrArea       0
BsmtFinSF1       0
BsmtFinSF2       0
BsmtUn

In [ ]:
train_x1 = pd.concat([train_num, train_cat], axis = 1)
test_x1 = pd.concat([test_num, test_cat], axis = 1) 

In [ ]:
encoder = OneHotEncoder(sparse = False, handle_unknown = 'ignore')
encoder.fit(train_x1)
train_x1 = pd.DataFrame(encoder.transform(train_x1), columns = encoder.get_feature_names_out())
test_x1 = pd.DataFrame(encoder.transform(test_x1), columns = encoder.get_feature_names_out())

/usr/local/lib/python3.9/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [ ]:
scaler = StandardScaler()
scaler.fit(train_x1)
train_x1 = pd.DataFrame(scaler.transform(train_x1), columns = train_x1.columns)
test_x1 = pd.DataFrame(scaler.transform(test_x1), columns = test_x1.columns)

In [ ]:
model1 = LinearRegression()
model2 = DecisionTreeRegressor(random_state = 69)
model3 = RandomForestRegressor(random_state = 69)
model4 = KNeighborsRegressor()

In [ ]:
model1.fit(train_x1, train_y)
pred1 = model1.predict(test_x1)

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error
mae1 = mean_absolute_error(test_y, pred1)

In [ ]:
mae1

24605.462528804826

In [ ]:
model2.fit(train_x1, train_y)
pred2 = model2.predict(test_x1)

from sklearn.metrics import mean_squared_error, mean_absolute_error
mae2 = mean_absolute_error(test_y, pred2)

mae2

32311.671232876713

In [ ]:
model3.fit(train_x1, train_y)
pred3 = model3.predict(test_x1)

from sklearn.metrics import mean_squared_error, mean_absolute_error
mae3 = mean_absolute_error(test_y, pred3)

mae3

22463.731561643835

In [ ]:
model4.fit(train_x1, train_y)
pred4 = model4.predict(test_x1)

from sklearn.metrics import mean_squared_error, mean_absolute_error
mae4 = mean_absolute_error(test_y, pred4)

mae4

81638.39287671233

Base model score is 24605.462 for Linear Regression, 32311.67 for Decision Tree Regressor, 22463.73 for Random Forest Regressor and 81638.39 for KNN Regressor. Now we will implement feature selection in order to increase the score!

In [ ]:
nan_feat = [cname for cname in data2.columns if data2[cname].isnull().sum() >= 1]

# TO FIND OUT THE % OF NAN 
for i in nan_feat:
    print(i, np.round(data2[i].isnull().mean(), 2))

LotFrontage 0.18
Alley 0.94
MasVnrType 0.01
MasVnrArea 0.01
BsmtQual 0.03
BsmtCond 0.03
BsmtExposure 0.03
BsmtFinType1 0.03
BsmtFinType2 0.03
Electrical 0.0
FireplaceQu 0.47
GarageType 0.06
GarageYrBlt 0.06
GarageFinish 0.06
GarageQual 0.06
GarageCond 0.06
PoolQC 1.0
Fence 0.81
MiscFeature 0.96


In [ ]:
cat_data = data2.select_dtypes(include = 'object')
pd.crosstab(cat_data['Street'], columns = 'counts', normalize = True)

col_0    counts
Street         
Grvl    0.00411
Pave    0.99589

In [ ]:
list1 = []
for i in cat_data.columns:
    list1.append((i, pd.crosstab(cat_data[i], columns = 'counts', normalize = True)))

In [ ]:
corr = data2.corr()
corr.style.background_gradient(cmap='coolwarm')

In [ ]:
low_var_list = ['Alley', 'YrSold', 'PoolQC', 'MiscFeature', 'MiscVal', 'GarageYrBlt', 'YearBuilt', 'MoSold', 
            '1stFlrSF', '2ndFlrSF', 'LotArea', 'YearRemodAdd', 'Street', 'Utilities', 'LandSlope', 
            'Condition2', 'RoofMatl', 'Heating', 'GarageCond']
data2.drop(low_var_list, axis = 1, inplace = True)

In [ ]:
x1 = data2.drop(['SalePrice'], axis = 1)
y1 = data2['SalePrice']

In [ ]:
train_x4, test_x4, train_y4, test_y4 = train_test_split(x1, y1, random_state = 69)

In [ ]:
train_num1 = train_x4.select_dtypes(include = 'number')
train_cat1 = train_x4.select_dtypes(include = 'object')

test_num1 = test_x4.select_dtypes(include = 'number')
test_cat1 = test_x4.select_dtypes(include = 'object')

In [ ]:
print('Missing values before imputation \n', train_num1.isnull().sum())
train_num1.fillna(train_num1.median(), inplace = True)
print('\n')
print('Missing values after imputation \n', train_num1.isnull().sum())


Missing values before imputation 
 MSSubClass         0
LotFrontage      188
OverallQual        0
OverallCond        0
MasVnrArea         6
BsmtFinSF1         0
BsmtFinSF2         0
BsmtUnfSF          0
TotalBsmtSF        0
LowQualFinSF       0
GrLivArea          0
BsmtFullBath       0
BsmtHalfBath       0
FullBath           0
HalfBath           0
BedroomAbvGr       0
KitchenAbvGr       0
TotRmsAbvGrd       0
Fireplaces         0
GarageCars         0
GarageArea         0
WoodDeckSF         0
OpenPorchSF        0
EnclosedPorch      0
3SsnPorch          0
ScreenPorch        0
PoolArea           0
dtype: int64


Missing values after imputation 
 MSSubClass       0
LotFrontage      0
OverallQual      0
OverallCond      0
MasVnrArea       0
BsmtFinSF1       0
BsmtFinSF2       0
BsmtUnfSF        0
TotalBsmtSF      0
LowQualFinSF     0
GrLivArea        0
BsmtFullBath     0
BsmtHalfBath     0
FullBath         0
HalfBath         0
BedroomAbvGr     0
KitchenAbvGr     0
TotRmsAbvGrd     0
Firepla

In [ ]:
print('Missing values before imputation \n', train_cat1.isnull().sum())
train_cat1.fillna(train_cat1.mode().loc[0], inplace = True)
print('\n')
print('Missing values after imputation \n', train_cat1.isnull().sum())

Missing values before imputation 
 MSZoning           0
LotShape           0
LandContour        0
LotConfig          0
Neighborhood       0
Condition1         0
BldgType           0
HouseStyle         0
RoofStyle          0
Exterior1st        0
Exterior2nd        0
MasVnrType         6
ExterQual          0
ExterCond          0
Foundation         0
BsmtQual          26
BsmtCond          26
BsmtExposure      27
BsmtFinType1      26
BsmtFinType2      27
HeatingQC          0
CentralAir         0
Electrical         0
KitchenQual        0
Functional         0
FireplaceQu      517
GarageType        62
GarageFinish      62
GarageQual        62
PavedDrive         0
Fence            884
SaleType           0
SaleCondition      0
dtype: int64


Missing values after imputation 
 MSZoning         0
LotShape         0
LandContour      0
LotConfig        0
Neighborhood     0
Condition1       0
BldgType         0
HouseStyle       0
RoofStyle        0
Exterior1st      0
Exterior2nd      0
MasVnrType    

In [ ]:
print('Missing values before imputation \n', test_cat1.isnull().sum())
test_cat1.fillna(train_cat1.mode().loc[0], inplace = True)
print('\n')
print('Missing values after imputation \n', test_cat1.isnull().sum())


Missing values before imputation 
 MSZoning           0
LotShape           0
LandContour        0
LotConfig          0
Neighborhood       0
Condition1         0
BldgType           0
HouseStyle         0
RoofStyle          0
Exterior1st        0
Exterior2nd        0
MasVnrType         2
ExterQual          0
ExterCond          0
Foundation         0
BsmtQual          11
BsmtCond          11
BsmtExposure      11
BsmtFinType1      11
BsmtFinType2      11
HeatingQC          0
CentralAir         0
Electrical         1
KitchenQual        0
Functional         0
FireplaceQu      173
GarageType        19
GarageFinish      19
GarageQual        19
PavedDrive         0
Fence            295
SaleType           0
SaleCondition      0
dtype: int64


Missing values after imputation 
 MSZoning         0
LotShape         0
LandContour      0
LotConfig        0
Neighborhood     0
Condition1       0
BldgType         0
HouseStyle       0
RoofStyle        0
Exterior1st      0
Exterior2nd      0
MasVnrType    

In [ ]:
print('Missing values before imputation \n', test_num1.isnull().sum())
test_num1.fillna(train_num1.median(), inplace = True)
print('\n')
print('Missing values after imputation \n', test_num1.isnull().sum())

Missing values before imputation 
 MSSubClass        0
LotFrontage      71
OverallQual       0
OverallCond       0
MasVnrArea        2
BsmtFinSF1        0
BsmtFinSF2        0
BsmtUnfSF         0
TotalBsmtSF       0
LowQualFinSF      0
GrLivArea         0
BsmtFullBath      0
BsmtHalfBath      0
FullBath          0
HalfBath          0
BedroomAbvGr      0
KitchenAbvGr      0
TotRmsAbvGrd      0
Fireplaces        0
GarageCars        0
GarageArea        0
WoodDeckSF        0
OpenPorchSF       0
EnclosedPorch     0
3SsnPorch         0
ScreenPorch       0
PoolArea          0
dtype: int64


Missing values after imputation 
 MSSubClass       0
LotFrontage      0
OverallQual      0
OverallCond      0
MasVnrArea       0
BsmtFinSF1       0
BsmtFinSF2       0
BsmtUnfSF        0
TotalBsmtSF      0
LowQualFinSF     0
GrLivArea        0
BsmtFullBath     0
BsmtHalfBath     0
FullBath         0
HalfBath         0
BedroomAbvGr     0
KitchenAbvGr     0
TotRmsAbvGrd     0
Fireplaces       0
GarageCars     

In [ ]:
train_x5 = pd.concat([train_num1, train_cat1], axis = 1)
test_x5 = pd.concat([test_num1, test_cat1], axis = 1) 

In [ ]:
encoder1 = OneHotEncoder(sparse = False, handle_unknown = 'ignore')
encoder1.fit(train_x5)
train_x5 = pd.DataFrame(encoder1.transform(train_x5), columns = encoder1.get_feature_names_out())
test_x5 = pd.DataFrame(encoder1.transform(test_x5), columns = encoder1.get_feature_names_out())

/usr/local/lib/python3.9/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [ ]:
scaler1 = StandardScaler()
scaler1.fit(train_x5)
train_x5 = pd.DataFrame(scaler1.transform(train_x5), columns = train_x5.columns)
test_x5 = pd.DataFrame(scaler1.transform(test_x5), columns = test_x5.columns)

In [ ]:
model5 = LinearRegression()
model6 = DecisionTreeRegressor(random_state = 69)
model7 = RandomForestRegressor(random_state = 69)
model8 = KNeighborsRegressor()

In [ ]:
model5.fit(train_x5, train_y4)
pred1 = model5.predict(test_x5)

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error
mae1 = mean_absolute_error(test_y4, pred1)

In [ ]:
mae1

2.5146330042665444e+16

In [ ]:
model6.fit(train_x5, train_y4)
pred2 = model6.predict(test_x5)

from sklearn.metrics import mean_squared_error, mean_absolute_error
mae2 = mean_absolute_error(test_y4, pred2)

mae2

31779.45205479452

In [ ]:
model7.fit(train_x5, train_y4)
pred3 = model7.predict(test_x5)

from sklearn.metrics import mean_squared_error, mean_absolute_error
mae3 = mean_absolute_error(test_y4, pred3)

mae3

22207.57888584475

In [ ]:
model8.fit(train_x5, train_y4)
pred4 = model8.predict(test_x5)

from sklearn.metrics import mean_squared_error, mean_absolute_error
mae4 = mean_absolute_error(test_y4, pred4)

mae4

60116.882739726025